userapi/src/main/java/fk/prof/userapi/cache/ClusteredProfileCache.java

```diff
@@ -0,0 +1,281 @@
+                    Pair.of(profileViewPair.first.result().getAggregatedSamples(traceName), (T) profileViewPair.second));
+            }
+
+            return doAsync(() -> getOrCreateView(profileName, traceName, profileViewType));
+        }
+
+        // profile not found, so initiate the profile loading and compute view after that
+        getAggregatedProfile(profileName).setHandler(ar -> {
+            if(ar.succeeded()) {
+                try {
+                    viewFuture.complete(getOrCreateView(profileName, traceName, profileViewType));
+                    return;
+                } catch (Exception e) {
+                    viewFuture.fail(e);
+                    return;
+                }
+            }
+            viewFuture.fail(ar.cause());
+        });
+
+        return viewFuture;
+    }
+
+    private Future<AggregatedProfileInfo> loadProfile(AggregatedProfileNamingStrategy profileName) {
+        Future<AggregatedProfileInfo> future = doAsync(() -> profileLoader.load(profileName));
+
+        cache.put(profileName, future);
+        future.setHandler(ar -> {
+            logger.info("Profile load complete. file: {}", profileName);
+            // load_profile might fail, regardless reinsert to take the new utilization into account.
+            cache.put(profileName, future);
+        });
+        return future;
+    }
+
+    /**
+     * Helper method to get the value from local cache. If the profile and corresponding future exists, then
+     * @param profileName
+     * @return Optional<AggregatedProfileInfo> if loaded profile is present.
+     *         Optional.empty() if the cache has no information.
+     * @throws ProfileLoadInProgressException if last request is still in progress
+     * @throws CachedProfileNotFoundException if last request failed to load
+     */
+    private Optional<AggregatedProfileInfo> getFromLocalCache(AggregatedProfileNamingStrategy profileName)
+        throws ProfileLoadInProgressException, CachedProfileNotFoundException {
+
+        Future<AggregatedProfileInfo> cachedProfileInfo = cache.get(profileName);
+
+        if (cachedProfileInfo == null) {
+            return Optional.empty();
+        }
+
+        if(!cachedProfileInfo.isComplete()) {
+            throw new ProfileLoadInProgressException(profileName);
+        }
+        else if (!cachedProfileInfo.succeeded()) {
+            throw new CachedProfileNotFoundException(cachedProfileInfo.cause());
+        }
+
+        return Optional.of(cachedProfileInfo.result());
+    }
+
+    /**
+     * Event handler for evicted profiles. Deletes the profile -> ip:port mapping fom shared store.
+     * @param onRemoval the on removal notification received on an entry removal
+     */
+    private void doCleanUpOnEviction(RemovalNotification<AggregatedProfileNamingStrategy, Future<AggregatedProfileInfo>> onRemoval) {
+        if(!RemovalCause.REPLACED.equals(onRemoval.getCause()) && onRemoval.wasEvicted()) {
```

Comment:
why is wasEvicted check necessary?

Reply:
So that the logic only runs when the eviction happens automatically and not on .remove(key) call.